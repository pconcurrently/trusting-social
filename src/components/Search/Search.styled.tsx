import styled from 'styled-components';

import { IApplicationTheme, devices } from '../../shared/themes';

interface ITheme {
  theme: IApplicationTheme;
}

export const StyledSearch = styled.div`
  width: 100%;
  max-width: 600px;
  position: relative;
  top: 120px;
  margin: 0 auto;

  @media ${devices.tablet} {
    top: 220px;
  }

  input {
    width: 100%;
    height: 60px;
    padding: 0 24px;
    background: ${({ theme }: ITheme) => theme.BASE_BACKGROUND};
    border: ${({ theme }: ITheme) => `1px solid ${theme.BORDER}`};
    box-sizing: border-box;
    border-radius: 4px 4px 0px 0px;
    color: ${({ theme }: ITheme) => theme.BASE_TEXT};
    font-size: 20px;
    line-height: 23px;
    font-weight: bold;
    position: absolute;
    z-index: 1;

    @media ${devices.tablet} {
      max-width: 600px;
    }

    &:focus {
      outline: none;
    }
  }
`;
