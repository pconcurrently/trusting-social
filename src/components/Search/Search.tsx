import React from 'react';
import { debounce } from 'lodash';

import { StyledSearch } from './Search.styled';
import { searchResults } from './search-results';

const DEBOUNCE_TIME = 300;

export const Search = () => {
  const [filteredResults, setFilteredResults] = React.useState<string[]>([]);
  const handleOnChange = (value: string) => {
    const filtered = searchResults.filter(
      r => r.toLowerCase().indexOf(value.toLowerCase()) !== -1 // Simple text search
    );
    setFilteredResults(filtered);
  };

  const deboucedHandleOnChange = debounce(handleOnChange, DEBOUNCE_TIME);

  return (
    <StyledSearch>
      <input
        type="text"
        placeholder="Try Vietnam"
        autoFocus
        onChange={
          (e: React.FormEvent<HTMLInputElement>) =>
            deboucedHandleOnChange(e.currentTarget.value) // this event won't be persisted after debouncing, pass directly the value
        }
      />
    </StyledSearch>
  );
};
