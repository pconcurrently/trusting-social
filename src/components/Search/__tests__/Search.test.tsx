import React from 'react';
import { shallow } from 'enzyme';
import debounce from 'lodash/debounce';
import { Search } from '../Search';

describe('Search', () => {
  const switchTheme = jest.fn();

  it('should render Search', () => {
    const container = shallow(<Search isDarkTheme={false} />);
    expect(container.debug()).toMatchSnapshot();
  });

  it('should render search-icon-light', () => {
    const container = shallow(<Search isDarkTheme={false} />);
    expect(container.find('img').prop('src')).toContain('search-icon-light');
  });

  it('should render search-icon-dark', () => {
    const container = shallow(<Search isDarkTheme={true} />);
    expect(container.find('img').prop('src')).toContain('search-icon-dark');
  });

  it('should fire deboucedHandleOnChange on onChange event', () => {
    const handleOnChange = jest.fn();
    const deboucedHandleOnChange = debounce(handleOnChange, 300);
    const event = { currentTarget: { value: 'random text' } };
    const container = shallow(<Search isDarkTheme={true} />);
    container.find('input').simulate('change', event);
    jest.useFakeTimers();
    setTimeout(() => {
      expect(deboucedHandleOnChange).toHaveBeenCalledTimes(1);
    }, 300);
    
  });
});
