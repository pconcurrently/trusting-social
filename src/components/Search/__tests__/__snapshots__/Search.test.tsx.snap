// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render Search 1`] = `
"<styled.div>
  <div className=\\"input-wrapper\\">
    <input type=\\"text\\" placeholder=\\"Try Vietnam\\" autoFocus={true} onChange={[Function: onChange]} />
    <img src=\\"search-icon-light.svg\\" alt=\\"Search\\" />
  </div>
  <ResultsList results={{...}} />
</styled.div>"
`;
